test/apps/iodemo/ucx_wrapper.cc

```diff
@@ -820,13 +824,57 @@ void UcxConnection::set_log_prefix(const struct sockaddr* saddr,
     memcpy(_log_prefix, ss.str().c_str(), length);
 }
 
-void UcxConnection::connect_common(ucp_ep_params_t &ep_params,
-                                   UcxCallback *callback)
+void UcxConnection::connect_tag(UcxCallback *callback)
 {
     const ucp_datatype_t dt_int = ucp_dt_make_contig(sizeof(uint32_t));
-    void *rreq;
     size_t recv_len;
 
+    // receive remote connection id
+    void *rreq = ucp_stream_recv_nb(_ep, &_remote_conn_id, 1, dt_int,
+                                    stream_recv_callback, &recv_len,
+                                    UCP_STREAM_RECV_FLAG_WAITALL);
+    if (rreq == NULL) {
+        established();
+    } else if (UCS_PTR_IS_ERR(rreq)) {
+        handle_connection_error(UCS_PTR_STATUS(rreq));
+    } else {
+        _ucx_status   = UCS_INPROGRESS;
+        _establish_cb = callback;
+    }
+
+    process_request("conn_id receive", rreq, callback);
+    if (UCS_PTR_IS_PTR(rreq)) {
+        _context._conns_in_progress.push_back(std::make_pair(
+                UcxContext::get_time() + _context._connect_timeout, this));
+    } else {
+        delete callback;
+    }
+
+    if (!UCS_PTR_IS_ERR(rreq)) {
+        // send local connection id
+        void *sreq = ucp_stream_send_nb(_ep, &_conn_id, 1, dt_int,
+                                        stream_send_callback, 0);
+        // we can don't check status here, in case if the endpoint is failed
+        // we should handle it in ep_params.err_handler.cb set above
+        if (UCS_PTR_IS_PTR(sreq)) {
+            ucp_request_free(sreq);
+        }
+    }
+}
+
+void UcxConnection::connect_am(UcxCallback *callback)
+{
+    // With AM use ep as a connection ID. AM receive callback provides
+    // reply ep, which can be used for finding a proper connection.
+    _conn_id = reinterpret_cast<uint64_t>(_ep);
+    established();
+    process_request("connect AM", NULL, callback);
+    delete callback;
```

Comment:
AFAIR in case of sen operation, the callback deletes itself. so in conn callback case, should be the same

Reply:
connection keeps poiter to the callback, I would like to avoid freed poiter there, so IMO better to delete and NULL'ize the poiter in the same place.

Reply:
after callback is called, set conn->callback = NULL, user should take care of its memory release